### DISCUSSION TITLE
How much of Skia is used in VCL applications?

### DISCUSSION THREAD

**Discussion by War3Evo** [skia4delphi/skia4delphi#178] [Q&A]
Hi,

If I do enable Skia for the VCL application.  If I don't explicitly use any Skia components or methods, does Skia speed up VCL in anyway?

**Comment by viniciusfbb** [comment]
No speed or quality improve in VCL. The app render replacement can be done just in FMX. In VCL, the majority of controls is drawn by the system operation, not by the application/VCL framework.

**Comment by woodie2** [comment]
so if I use Skia for canvas drawing on TBitmap, it is not faster then the basic canvas drawing?

**Reply by viniciusfbb** [reply]
Drawings in TBitmap using Skia currently use raster (CPU), so similar performance is expected, maybe a little faster as there are several internal optimizations in Skia, but in Vcl the first drawing can have a small delay because it converts the pixelformat and the alpha format for pf32 and afPremultiplied (this might be faster if you set this manually before Bitmap.SetSize and before Bitmap.SkiaDraw, [like this](https://github.com/skia4delphi/skia4delphi/blob/ca0395079bb5723556aa1c6e0a8a4464a0016964/Source/VCL/Skia.Vcl.pas#L5848-L5855)). If your goal is to paint on a bitmap and then paint on a control, it might be better to directly use a TSkPaintBox control, as it already has some optimizations for that.

In the future we intend to change TBitmap drawings to use GPU internally, but there are many other demands from users before that.

**Reply by War3Evo** [reply]
So, converting a VCL project to FMX then enabling Skia4Delphi will or won't have a performance increase?

**Reply by viniciusfbb** [reply]
Will increase the performance.

**Reply by woodie2** [reply]
> In the future we intend to change TBitmap drawings to use GPU internally, but there are many other demands from users before that.

TBitmap with GPU drawing would be an main improvements.
and if you could also draw polygons with holes, it would be perfect

**Reply by viniciusfbb** [reply]
In fact, at least on my notebook, and in most desktop app draws, there isn't much difference in performance. There is a bigger difference in some specific draws, like in shaders. But for example, our "FmxFPS" benchmark project, that render simples controls like buttons, labels, images, rectangles and circles, on my notebook scores 82 fps rendering with GPU and 72 fps using CPU, that is, it has good performance in both.

**Reply by viniciusfbb** [reply]
> and if you could also draw polygons with holes, it would be perfect

it is already possible. You need to clip the holes before drawing the polygons.
